moneriote/dns/cloudflare.py

```diff
@@ -0,0 +1,71 @@
+from moneriote.dns import DnsProvider
+from moneriote.rpc import RpcNode, RpcNodeList
+from moneriote.utils import log_err, log_msg, random_user_agent, make_json_request
+
+
+class Cloudflare(DnsProvider):
+    def __init__(self, **kwargs):
+        super(Cloudflare, self).__init__(**kwargs)
+
+        self.headers = {
+            'Content-Type': 'application/json',
+            'X-Auth-Email': kwargs['api_email'],
+            'X-Auth-Key': kwargs['api_key'],
+            'User-Agent': random_user_agent()
+        }
+        self.api_base = 'https://api.cloudflare.com/client/v4/zones'
+        self.zone_id = None
+
+        # zone_id is required and will be detected via Cloudflare API
+        if not self.zone_id:
+            log_msg('Determining zone_id; looking for \'%s\'' % self.domain_name)
+            result = make_json_request(url=self.api_base, headers=self.headers)
+            try:
+                zones = result.get('result')
+                self.zone_id = next(zone.get('id') for zone in zones if zone.get('name') == self.domain_name)
+            except StopIteration:
+                log_err('could not determine zone_id. Is your Cloudflare domain correct?', fatal=True)
+            log_msg('Cloudflare zone_id \'%s\' matched to \'%s\'' % (self.zone_id, self.domain_name))
+
+    def get_records(self):
+        nodes = RpcNodeList()
+        log_msg('Fetching existing record(s) (%s.%s)' % (self.subdomain_name, self.domain_name))
+
+        result = make_json_request('%s/%s/dns_records/' % (self.api_base, self.zone_id), headers=self.headers)
+        records = result.get('result')
```

Comment:
Got an exception here
```
[2018-09-27 22:50] GET: https://api.cloudflare.com/client/v4/zones/xxx/dns_records/
[2018-09-27 22:50] Error (https://api.cloudflare.com/client/v4/zones/xxx/dns_records/): 500 Server Error: Internal Server Error for url: https://api.cloudflare.com/client/v4/zones/xxx/dns_records/
Traceback (most recent call last):
  File "C:\Users\walass\Documents\moneriote-python\venv\Scripts\moneriote-script.py", line 11, in <module>    load_entry_point('moneriote', 'console_scripts', 'moneriote')()
  File "C:\Users\walass\Documents\moneriote-python\venv\lib\site-packages\click-7.0-py3.6.egg\click\core.py", line 764, in __call__    return self.main(*args, **kwargs)
  File "C:\Users\walass\Documents\moneriote-python\venv\lib\site-packages\click-7.0-py3.6.egg\click\core.py", line 717, in main    rv = self.invoke(ctx)
  File "C:\Users\walass\Documents\moneriote-python\venv\lib\site-packages\click-7.0-py3.6.egg\click\core.py", line 956, in invoke    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\walass\Documents\moneriote-python\venv\lib\site-packages\click-7.0-py3.6.egg\click\core.py", line 555, in invoke    return callback(*args, **kwargs)
  File "c:\users\walass\documents\moneriote-python\moneriote\main.py", line 88,in cli    mon.main()
  File "c:\users\walass\documents\moneriote-python\moneriote\moneriote.py", line 74, in main    dns_nodes = self.dns_provider.get_records()
  File "c:\users\walass\documents\moneriote-python\moneriote\dns\cloudflare.py", line 35, in get_records    records = result.get('result')
AttributeError: 'NoneType' object has no attribute 'get'
```

Reply:
Different conditions log:
```
[2018-09-27 23:00] Fetching existing record(s) (testopennode.xmr-tw.org)
[2018-09-27 23:00] GET: https://api.cloudflare.com/client/v4/zones/xxxxx/dns_records/
[2018-09-27 23:00] Error (https://api.cloudflare.com/client/v4/zones/xxxxx/dns_records/): HTTPSConnectionPool(host='api.cloudflare.com', port=443): Read timed out. (read timeout=5)
```
```
[2018-09-27 23:01] Fetching existing record(s) (testopennode.xmr-tw.org)
[2018-09-27 23:01] GET: https://api.cloudflare.com/client/v4/zones/xxxxx/dns_records/
[2018-09-27 23:01] Record insertion: 71.127.156.63
[2018-09-27 23:01] POST: https://api.cloudflare.com/client/v4/zones/xxxxx/dns_records
[2018-09-27 23:01] Record insertion: 188.194.32.122
[2018-09-27 23:01] POST: https://api.cloudflare.com/client/v4/zones/xxxxx/dns_records
[2018-09-27 23:01] Record insertion: 142.93.216.168
[2018-09-27 23:01] POST: https://api.cloudflare.com/client/v4/zones/xxxxx/dns_records
[2018-09-27 23:01] Record insertion: 159.203.35.198
[2018-09-27 23:01] POST: https://api.cloudflare.com/client/v4/zones/xxxxx/dns_records
[2018-09-27 23:01] Record insertion: 95.90.66.92
[2018-09-27 23:01] POST: https://api.cloudflare.com/client/v4/zones/xxxxx/dns_records
[2018-09-27 23:01] Error (https://api.cloudflare.com/client/v4/zones/xxxxx/dns_records): 400 Client Error: Bad Request for url: https://api.cloudflare.com/client/v4/zones/xxxxx/dns_records
```

Reply:
I figured out this issue is caused by cloudflare API only returning 20 records by default. But my DNS records is way more than that so script couldn't get any valid record in `get_records` function. This makes issue of record deletion. [Filter records name directly by CF API](https://api.cloudflare.com/#dns-records-for-a-zone-list-dns-records) might be the better way than current method: `if record.get('type') != 'A' or record.get('name') != self.fulldomain_name:`

Reply:
I tried this and it works well.
``` python
log_msg('Fetching existing record(s) (%s.%s)' % (self.subdomain_name, self.domain_name))
result = make_json_request('%s/%s/dns_records/?type=A&name=%s.%s' % (self.api_base, self.zone_id,self.subdomain_name, self.domain_name), headers=self.headers)
records = result.get('result')
```

Reply:
oh, very nice. Thanks!